### ISSUE TITLE
Bug in Cross-Validation estimation

### ISSUE THREAD

**Issue by davidwaroquiers** [microsoft/FLAML#52]
Dear all,

I have been trying FLAML for a few days now and I believe I stumbled across a bug in the evaluation of the model when using cross-validation (eval_method="cv").

I believe that there is only the last fold that is taken into account in function evaluate_model_CV (ml.py). The list of validation scores (val_loss_list) is only updated with the current fold's validation score for the last fold or when the budget is not anymore sufficient. In any case, the val_loss_list only contains one item in all cases. Moreover, what is appended to the list is not the validation score of the current fold, but the mean of the validation scores of the first "valid_fold_num" folds.

I would suggest the following to replace lines 220--226 in ml.py:

```
        val_loss_list.append(val_loss_i)
        if valid_fold_num == n:
            total_val_loss = valid_fold_num = 0
        elif time.time() - start_time >= budget:
            break
    val_loss = np.max(val_loss_list)
```

One might also consider changing (or make some options) for the last line in the above snippet. Indeed, here the maximum of the validation scores of each fold is taken. Another commonly used way is to take the average of the validation scores of each fold. This could be an option for the user but it is not a bug per se. I am also ok keeping the max of all validation scores as it is now. (note that basically, the current situation is using the mean value of the different folds, as it is taking the total_val_loss divided by the number of folds).

Best

David

**Comment by sonichi** [comment]
Hi David, thanks for the comment. It doesn't produce unwanted result but I understand why it looks overcomplicated in the current form. With `n_repeats=1` in `kf`, `val_loss_list` only contains one element which is the average of loss from `valid_fold_num` folds. So we don't really need the list as of now. Using the list, however, allows setting `n_repeats` larger than 1.

Please let us know if you find other ways of cross validation are needed in your application.

**Comment by davidwaroquiers** [comment]
Hi,

Thank you for your answer. Ok indeed, I see. What I wanted is to have access to the different validation scores of each fold and it was difficult (actually not possible without changing a little bit the code of course) to do that with the current implementation.

Best,

David

**Comment by sonichi** [comment]
"to have access to the different validation scores of each fold", do you need the scores per fold to be logged? Otherwise they are not going to be accessible from outside.

**Comment by sonichi** [comment]
@davidwaroquiers Since this is not really a bug I will close this issue now. If you need the scores per fold to be logged please create a separate issue for that feature request.
